sunpy/coordinates/spice.py

```diff
@@ -0,0 +1,170 @@
+"""
+Experimental module to use the SkyCoord API for SPICE computations
+
+.. warning::
+    This module is under development, so may be subject to significant change.
+
+The `SPICE <https://naif.jpl.nasa.gov/naif/>`_ observation geometry
+information system is being increasingly used by space missions to describe the
+locations of spacecraft and the time-varying orientations of reference frames.
+While the `~spiceypy.spiceypy` package provides a Python interface for
+performing SPICE computations, its API is very different from that of
+`~astropy.coordinates.SkyCoord`.
+
+This module "wraps" `~spiceypy.spiceypy` functionality so that relevant SPICE
+computations can be accessed using the `~astropy.coordinates.SkyCoord` API.
+When loading a set of kernels, a frame class and corresponding transformations
+are created for each SPICE frame.  One can also query the location of a body
+as computed via SPICE.
+
+.. note::
+    This module requires the optional dependency `~spiceypy.spiceypy` to be
+    installed.
+
+Notes
+-----
+* All transformations from one SPICE frame to another SPICE frame go through
+  `~astropy.coordinates.ICRS` as the intermediate frame, even if the origin
+  shift to/from the solar-system barycenter is unnatural.
+* There is currently no support for time arrays.
+"""
+
+try:
+    import spiceypy
+except ImportError:
+    raise ImportError("This module requires the optional dependency `spiceypy`.")
+
+import astropy.units as u
+from astropy.coordinates import ICRS, SkyCoord, frame_transform_graph
+from astropy.coordinates.representation import CartesianRepresentation
+from astropy.coordinates.transformations import FunctionTransformWithFiniteDifference
+from astropy.time import Time
+
+from sunpy import log
+from sunpy.coordinates import SunPyBaseCoordinateFrame
+from sunpy.time import parse_time
+from sunpy.time.time import _variables_for_parse_time_docstring
+from sunpy.util.decorators import add_common_docstring
+
+__all__ = ['get_body', 'initialize']
+
+
+# Note that this epoch is very slightly different from the typical definition of J2000.0 (in TT)
+_ET_REF_EPOCH = Time('J2000', scale='tdb')
+
+_CLASS_TYPES = ['', 'inertial', 'PCK', 'CK', 'TK', 'dynamic', 'switch']
+
+
+# Defined for future use
+class _SpiceBaseCoordinateFrame(SunPyBaseCoordinateFrame):
+    pass
+
+
+def _convert_to_et(time):
+    return (time - _ET_REF_EPOCH).to_value('s')
+
+
+def _install_frame(frame_id):
+    frame_name = spiceypy.frmnam(frame_id)
+    # TODO: Sanitize/escape the frame name of special characters
```

Comment:
Do we need to sanitise it to a valid Python identifier? If so we do something similar in net for the attributes on `Attr`: https://github.com/sunpy/sunpy/blob/main/sunpy/net/attr.py#L241 perhaps we could spin that out into a util function?

Reply:
The allowed characters for a SPICE frame name are alphanumeric, underscore, plus, and minus.  I don't know what to do with plus or minus characters.  Presumably the least bad choice would be to replace them with underscores, but collisions would be bad.  I'm keeping SPICE frame names in all uppercase, so I could replace them with lowercase p/n characters?

Reply:
> I'm keeping SPICE frame names in all uppercase, so I could replace them with lowercase p/n characters?

This is what I've done